File: okbinder/src/main/java/cn/thens/okbinder/OkBinder.java
package cn.thens.okbinder;

import android.os.Binder;
import android.os.IBinder;
import android.util.Base64;
import android.util.Log;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;
import java.lang.reflect.Method;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@SuppressWarnings({"ConstantConditions", "unchecked"})
public final class OkBinder {
    private static final String TAG = "@OkBinder";

    @Retention(RetentionPolicy.RUNTIME)
    @Target(ElementType.TYPE)
    public @interface Interface {
    }

    public static Binder create(Object remoteObject) {
        Class<?> okBinderInterface = getOkBinderInterface(remoteObject);
        require(okBinderInterface != null,
                "remote object must implement only one interface with @OkBinder.Interface annotation");
        return create(remoteObject, (Class<Object>) okBinderInterface);
    }

    public static <T> Binder create(T remoteObject, Class<T> okBinderInterface) {
        return getFactory(okBinderInterface).newBinder(remoteObject, okBinderInterface);
    }

    public static <T> T proxy(IBinder binder, Class<T> serviceClass) {
        return (T) getFactory(serviceClass).newProxy(binder, serviceClass);
    }

    private static final Map<Class<?>, OkBinderFactory> factories = new HashMap<>();
    private static final OkBinderFactory defaultFactory = new ReflectionFactory();

    private static OkBinderFactory getFactory(Class<?> serviceClass) {
        require(isOkBinderInterface(serviceClass),
                "service class must be an interface with @OkBinder.Interface annotation");
        OkBinderFactory factory = factories.get(serviceClass);
        if (factory != null) {
            return factory;
        }
        String factoryClassName = serviceClass.getName() + "_OkBinderFactory";
        try {
            ClassLoader classLoader = serviceClass.getClassLoader();
            Class<?> factoryClass = classLoader.loadClass(factoryClassName);
            factory = (OkBinderFactory) factoryClass.newInstance();
        } catch (Throwable e) {
            Log.w(TAG, "load factory class '" + factoryClassName + "' failed, " +
                    "use default factory as fallback");
        }
        if (factory == null) {
            factory = defaultFactory;
        }
        factories.put(serviceClass, factory);
        return factory;
    }

    private static void require(boolean value, String message) {
        if (!value) throw new IllegalArgumentException(message);
    }

    static boolean isOkBinderInterface(Class<?> cls) {
        return cls.isInterface() && cls.isAnnotationPresent(Interface.class);
    }

    static Class<?> getOkBinderInterface(Object object) {
        Class<?>[] interfaces = object.getClass().getInterfaces();
        List<Class<?>> okBinderInterfaces = new ArrayList<>();
        for (Class<?> anInterface : interfaces) {
            if (isOkBinderInterface(anInterface)) {
                okBinderInterfaces.add(anInterface);
            }
        }
        if (okBinderInterfaces.size() != 1) return null;
        return okBinderInterfaces.get(0);
    }

    public static String getMethodId(Method method) {
        StringBuilder methodSignatureBuilder = new StringBuilder();
        for (Class<?> parameterType : method.getParameterTypes()) {
            methodSignatureBuilder.append(",").append(parameterType.getName());
        }
        String methodSignature = methodSignatureBuilder.toString();
        try {
            byte[] md5 = MessageDigest.getInstance("MD5").digest(methodSignature.getBytes());
            return method.getName() + ":" + Base64.encodeToString(md5, Base64.NO_WRAP);
        } catch (Throwable e) {
            throw ErrorUtils.wrap(e);
        }
    }
}

File: okbinder/src/main/java/cn/thens/okbinder/OkBinder.kt
package cn.thens.okbinder

import android.os.Binder
import android.os.IBinder
import android.os.Parcel
import android.util.Base64
import android.util.Log
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.lang.reflect.Proxy
import java.security.MessageDigest
import java.util.concurrent.ConcurrentHashMap

class OkBinder(private val remoteObject: Any) : Binder() {
    private val remoteMethods = ConcurrentHashMap<String, Method>()
    private val descriptor: String

    init {
        val remoteInterface = (remoteObject.javaClass.interfaces ?: emptyArray())
            .filter { it.isOkBinderInterface() }
            .also { require(it.size == 1) { "remote object must implement only one interface with @OkBinder.Interface annotation" } }
            .first()
        for (method in remoteInterface.declaredMethods) {
            if (method.isBridge) continue
            remoteMethods[getMethodId(method)] = method
        }
        descriptor = remoteInterface.name
        attachInterface({ this }, descriptor)
    }

    override fun onTransact(code: Int, data: Parcel, reply: Parcel?, flags: Int): Boolean {
        requireNotNull(reply)
        if (code == IBinder.FIRST_CALL_TRANSACTION) {
            try {
                data.enforceInterface(descriptor)
                val remoteMethod = remoteMethods.getValue(data.readString()!!)
                val paramTypes = remoteMethod.parameterTypes
                val result = if (paramTypes.isNullOrEmpty()) {
                    remoteMethod.invoke(remoteObject)
                } else {
                    val classLoader = remoteObject.javaClass.classLoader
                    remoteMethod.invoke(remoteObject, *run {
                        Array(paramTypes.size) { adaptToProxy(data.readValue(classLoader), paramTypes[it]) }
                    })
                }
                reply.writeNoException()
                if (result != null) {
                    reply.writeInt(1)
                    reply.writeValue(adaptToBinder(result, remoteMethod.returnType))
                } else {
                    reply.writeInt(0)
                }
                return true
            } catch (e: Exception) {
                val cause = e.cause.takeIf { e is InvocationTargetException } ?: e
                Log.e("@OkBinder", "from Service", cause)
                throw cause
            }
        }
        return super.onTransact(code, data, reply, flags)
    }

    @Target(AnnotationTarget.CLASS)
    annotation class Interface

    companion object {
        @Suppress("UNCHECKED_CAST")
        fun <T> proxy(binder: IBinder, serviceClass: Class<T>): T {
            require(serviceClass.isOkBinderInterface()) { "class must be an interface with @OkBinder.Interface annotation" }
            if (binder is OkBinder) return binder.remoteObject as T
            val classLoader = serviceClass.classLoader
            return serviceClass.cast(Proxy.newProxyInstance(classLoader, arrayOf(serviceClass)) { _, method, args ->
                if (!binder.isBinderAlive) throw RuntimeException("binder has died")
                val data = Parcel.obtain()
                val reply = Parcel.obtain()
                var result: Any? = null
                try {
                    data.writeInterfaceToken(serviceClass.name)
                    data.writeString(getMethodId(method))
                    args?.forEachIndexed { index, arg ->
                        data.writeValue(adaptToBinder(arg, method.parameterTypes[index]))
                    }
                    val flags = if (method.returnType == Void.TYPE) IBinder.FLAG_ONEWAY else 0
                    binder.transact(IBinder.FIRST_CALL_TRANSACTION, data, reply, flags)
                    reply.readException()
                    if (reply.readInt() != 0) {
                        result = adaptToProxy(reply.readValue(classLoader)!!, method.returnType)
                    }
                } finally {
                    reply.recycle()
                    data.recycle()
                }
                result
            })!!
        }

        private fun adaptToBinder(value: Any?, type: Class<*>): Any? {
            return if (value != null && type.isOkBinderInterface()) OkBinder(value) else value
        }

        private fun adaptToProxy(value: Any?, type: Class<*>): Any? {
            return if (value != null && type.isOkBinderInterface()) proxy(value as IBinder, type) else value
        }

        private fun getMethodId(method: Method): String {
            val params = method.parameterTypes.joinToString(",") { it.name }
            val methodSignature = method.declaringClass.name + "." + method.name + "@" + params
            val md5 = MessageDigest.getInstance("MD5").digest(methodSignature.toByteArray())
            return Base64.encodeToString(md5, Base64.NO_WRAP)
        }

        private fun Class<*>.isOkBinderInterface(): Boolean {
            return isInterface && isAnnotationPresent(Interface::class.java)
        }
    }
}